File: src/Calendrie.Testing/CSharpTests/BoundedBelowScopeTests.cs
// SPDX-License-Identifier: BSD-3-Clause
// Copyright (c) Tran Ngoc Bich. All rights reserved.

namespace Calendrie.Testing.CSharpTests;

using Calendrie.Core.Schemas;
using Calendrie.Hemerology;
using Calendrie.Testing.Data;
using Calendrie.Testing.Data.Schemas;

public partial class BoundedBelowScopeTests : CalendricalDataConsumer<GregorianDataSet>
{
    private const int FirstYear = 3;
    private const int FirstMonth = 4;
    private const int FirstDay = 5;

    private static readonly GregorianSchema s_Schema = new();

    [Fact]
    public static void Create_NullSchema() =>
        AssertEx.ThrowsAnexn("schema",
            () => BoundedBelowScope.Create(null!, DayZero.NewStyle, new(3, 4, 5), 9999));

    [Fact]
    public static void Create_InvalidMinYear() =>
        Assert.Throws<ArgumentOutOfRangeException>(
            () => BoundedBelowScope.Create(
                s_Schema, DayZero.NewStyle, new(s_Schema.SupportedYears.Min - 1, 1, 1), 9999));

#if ENABLE_GENERIC_FACTORIES
    [Fact]
    public static void Create_InvalidMinYear_Generic() =>
        Assert.Throws<ArgumentOutOfRangeException>(
            () => BoundedBelowScope.Create<GregorianSchema>(
                DayZero.NewStyle, new(s_Schema.SupportedYears.Min - 1, 1, 1), 9999));
#endif

    [Fact]
    public static void Create_InvalidMaxYear() =>
        Assert.Throws<ArgumentOutOfRangeException>(
            () => BoundedBelowScope.Create(
                s_Schema, DayZero.NewStyle, new(s_Schema.SupportedYears.Max + 1, 1, 1), 9999));

#if ENABLE_GENERIC_FACTORIES
    [Fact]
    public static void Create_InvalidMaxYear_Generic() =>
        Assert.Throws<ArgumentOutOfRangeException>(
            () => BoundedBelowScope.Create<GregorianSchema>(
                DayZero.NewStyle, new(s_Schema.SupportedYears.Max + 1, 1, 1), 9999));
#endif

    [Theory, MemberData(nameof(InvalidMonthFieldData))]
    public static void Create_InvalidMonth(int y, int m) =>
        Assert.Throws<ArgumentOutOfRangeException>(
            () => BoundedBelowScope.Create(s_Schema, DayZero.NewStyle, new(y, m, 1), 9999));

#if ENABLE_GENERIC_FACTORIES
    [Theory, MemberData(nameof(InvalidMonthFieldData))]
    public static void Create_InvalidMonth_Generic(int y, int m) =>
        Assert.Throws<ArgumentOutOfRangeException>(
            () => BoundedBelowScope.Create<GregorianSchema>(DayZero.NewStyle, new(y, m, 1), 9999));
#endif

    [Theory, MemberData(nameof(InvalidDayFieldData))]
    public static void Create_InvalidDay(int y, int m, int d) =>
        Assert.Throws<ArgumentOutOfRangeException>(
            () => BoundedBelowScope.Create(s_Schema, DayZero.NewStyle, new(y, m, d), 9999));

#if ENABLE_GENERIC_FACTORIES
    [Theory, MemberData(nameof(InvalidDayFieldData))]
    public static void Create_InvalidDay_Generic(int y, int m, int d) =>
        Assert.Throws<ArgumentOutOfRangeException>(
            () => BoundedBelowScope.Create<GregorianSchema>(DayZero.NewStyle, new(y, m, d), 9999));
#endif

    [Theory(Skip = "MinDateParts cannot be the start of a year."), MemberData(nameof(DateInfoData))]
    public void Create(DateInfo info)
    {
        var (y, m, d, doy) = info;
        // Act
        var scope = BoundedBelowScope.Create(s_Schema, DayZero.NewStyle, new(y, m, d), 9999);
        var minDate = scope.Segment.MinMaxDateParts.LowerValue;
        var minOrdinalDate = scope.Segment.MinMaxOrdinalParts.LowerValue;
        // Assert
        Assert.NotNull(scope);
        Assert.Equal(y, scope.Segment.SupportedYears.Min);
        Assert.Equal(y, minDate.Year);
        Assert.Equal(m, minDate.Month);
        Assert.Equal(d, minDate.Day);
        Assert.Equal(9999, scope.Segment.SupportedYears.Max);
        Assert.Equal(doy, minOrdinalDate.DayOfYear);
    }

    [Fact]
    public static void ValidateYearMonth()
    {
        var scope = BoundedBelowScope.Create(
            s_Schema, DayZero.NewStyle, new(FirstYear, FirstMonth, FirstDay), 9999);
        // Act
        scope.ValidateYearMonth(FirstYear, FirstMonth);
        AssertEx.ThrowsAoorexn("month", () => scope.ValidateYearMonth(FirstYear, FirstMonth - 1));
    }

    [Fact]
    public static void ValidateYearMonthDay()
    {
        var scope = BoundedBelowScope.Create(
            s_Schema, DayZero.NewStyle, new(FirstYear, FirstMonth, FirstDay), 9999);
        // Act
        scope.ValidateYearMonthDay(FirstYear, FirstMonth, FirstDay);
        scope.ValidateYearMonthDay(FirstYear, FirstMonth, FirstDay + 1);
        scope.ValidateYearMonthDay(FirstYear, FirstMonth + 1, 1);
        AssertEx.ThrowsAoorexn("month", () => scope.ValidateYearMonthDay(FirstYear, FirstMonth - 1, FirstDay));
        AssertEx.ThrowsAoorexn("day", () => scope.ValidateYearMonthDay(FirstYear, FirstMonth, FirstDay - 1));
    }

    [Fact]
    public static void ValidateOrdinal()
    {
        int firstDayOfYear = s_Schema.CountDaysInYearBeforeMonth(FirstYear, FirstMonth) + FirstDay;
        var scope = BoundedBelowScope.Create(
            s_Schema, DayZero.NewStyle, new(FirstYear, FirstMonth, FirstDay), 9999);
        // Act
        scope.ValidateOrdinal(FirstYear, firstDayOfYear);
        AssertEx.ThrowsAoorexn("dayOfYear", () => scope.ValidateOrdinal(FirstYear, firstDayOfYear - 1));
    }
}

File: src/Calendrie.Tests/Tests/Hemerology/BoundedBelowScopeTests.fs
﻿// SPDX-License-Identifier: BSD-3-Clause
// Copyright (c) Tran Ngoc Bich. All rights reserved.

module Calendrie.Tests.Hemerology.BoundedBelowScopeTests

open Calendrie
open Calendrie.Core.Schemas
open Calendrie.Hemerology
open Calendrie.Testing

open Xunit

let private sch = new GregorianSchema()
let minYear = sch.SupportedYears.Min
let maxYear = sch.SupportedYears.Max

module Factories =
    [<Fact>]
    let ``Create() throws for null schema`` () =
        nullExn "schema" (fun () -> BoundedBelowScope.Create(null, DayZero.NewStyle, new DateParts(1, 1, 1), 2))

    [<Fact>]
    let ``Create() throws for an invalid date`` () =
        let sch = new GregorianSchema()

        outOfRangeExn "value" (fun () -> BoundedBelowScope.Create(sch, DayZero.NewStyle, new DateParts(minYear - 1, 12, 1), maxYear))
        outOfRangeExn "value" (fun () -> BoundedBelowScope.Create(sch, DayZero.NewStyle, new DateParts(maxYear + 1, 12, 1), maxYear))
        outOfRangeExn "value" (fun () -> BoundedBelowScope.Create(sch, DayZero.NewStyle, new DateParts(1, 0, 1), 2))
        outOfRangeExn "value" (fun () -> BoundedBelowScope.Create(sch, DayZero.NewStyle, new DateParts(1, 13, 1), 2))
        outOfRangeExn "value" (fun () -> BoundedBelowScope.Create(sch, DayZero.NewStyle, new DateParts(1, 1, 0), 2))
        outOfRangeExn "value" (fun () -> BoundedBelowScope.Create(sch, DayZero.NewStyle, new DateParts(1, 1, 32), 2))

    [<Fact>]
    let ``Create() throws for an invalid max year`` () =
        outOfRangeExn "year" (fun () -> BoundedBelowScope.Create(new GregorianSchema(), DayZero.NewStyle, new DateParts(1, 12, 1), maxYear + 1))

    [<Fact>]
    let ``Create()`` () =
        let parts = new DateParts(5, 3, 13)
        let maxYear = 15
        let scope = BoundedBelowScope.Create(new GregorianSchema(), DayZero.NewStyle, parts, maxYear)
        let seg = scope.Segment

        seg.MinIsStartOfYear |> nok
        seg.MaxIsEndOfYear   |> ok
        seg.IsComplete       |> nok

        seg.MinMaxDateParts.LowerValue === parts
        seg.MinMaxDateParts.UpperValue === new DateParts(maxYear, 12, 31)
        seg.SupportedYears.Max === maxYear

    [<Fact>]
    let ``Create() throws when the min date is the start of a year`` () =
        let parts = new DateParts(5, 1, 1)
        let maxYear = 15

        argExn "segment" (fun () -> BoundedBelowScope.Create(new GregorianSchema(), DayZero.NewStyle, parts, maxYear))

    [<Fact>]
    let ``StartingAt()`` () =
        let scope = BoundedBelowScope.StartingAt(new GregorianSchema(), DayZero.NewStyle, new DateParts(1, 12, 1))
        let seg = scope.Segment

        seg.MinIsStartOfYear |> nok
        seg.MaxIsEndOfYear   |> ok
        seg.IsComplete       |> nok

        seg.SupportedYears.Max === maxYear
